### PR TITLE
2810 api being able to refer parameters values from another model

### PR DESCRIPTION
**Pull request by barondim**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [x] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [x] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

**Comment by barondim** [comment]
- [ ] Doxygen doc not finished

File: dynawo/sources/API/PAR/CMakeLists.txt [hunk 1]
```diff
@@ -22,7 +22,6 @@ set(API_PAR_SOURCES
     PARParameter.cpp
     PARParameterFactory.cpp
     PARReference.cpp
-    PARReferenceFactory.cpp
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/CMakeLists.txt:25]: Please do not remove this as it is used in DFL

File: dynawo/sources/API/PAR/PARReference.cpp [hunk 2]
```diff
@@ -27,7 +27,14 @@ using std::string;
 
 namespace parameters {
 
-Reference::Reference(const string& name) : type_(""), name_(name), origName_(""), componentId_("") {}
+Reference::Reference(const string& name, OriginData origData) :
+  type_(""),
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/PARReference.cpp:31]: useless (same for all strings)

File: dynawo/sources/API/PAR/test/TestReference.cpp [hunk 3]
```diff
@@ -32,28 +31,41 @@ namespace parameters {
...
+  shared_ptr<Reference> refencePar = boost::shared_ptr<Reference>(new Reference("standardRef"));
 
   // Test default attributes
   ASSERT_EQ(refencePar->getType(), "");
   ASSERT_EQ(refencePar->getName(), "standardRef");
+  ASSERT_EQ(refencePar->getOrigData(), Reference::OriginData::IIDM);
+  ASSERT_EQ(refencePar->getOrigDataStr(), "IIDM");
   ASSERT_EQ(refencePar->getOrigName(), "");
   ASSERT_EQ(refencePar->getComponentId(), "");
+  ASSERT_EQ(refencePar->getParId(), "");
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/test/TestReference.cpp:43]: I would propose ASSERT_TRUE(refencePar->getParId().empty()

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 4]
```diff
@@ -234,6 +230,7 @@ DynamicData::associateParameters() {
       case Model::BLACK_BOX_MODEL: {
         shared_ptr<BlackBoxModel> bbm = dynamic_pointer_cast<BlackBoxModel>(model);
         shared_ptr<ParametersSet> modelSet = getParametersSet(bbm->getId(), bbm->getParFile(), bbm->getParId());
+        addReferencesFromParFilesInParSet(bbm, modelSet);
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:233]: Should be done in ModelicaModels also

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 5]
```diff
@@ -274,6 +271,36 @@ DynamicData::getParametersSet(const string& modelId, const string& parFile, cons
...
 
+void
+DynamicData::addReferencesFromParFilesInParSet(shared_ptr<BlackBoxModel> bbm, shared_ptr<ParametersSet> modelSet) const {
+  if (modelSet == nullptr)
+    return;
+  const std::vector<std::string> parReferences = modelSet->getParReferences();
+  for (const std::string parRefName : parReferences) {
+    const shared_ptr<Reference> parRef = modelSet->getReference(parRefName);
+    std::string parFile;
+    if (parRef->getParFile() == "") {
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:282]: if (parRef->getParFile().empty()) is more efficient

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 6]
```diff
@@ -274,6 +271,36 @@ DynamicData::getParametersSet(const string& modelId, const string& parFile, cons
...
+      parFile = bbm->getParFile();
+    } else {
+      parFile = parRef->getParFile();
+    }
+    std::string parFilePath = canonical(parFile, rootDirectory_);
+    parameters::XmlImporter parametersImporter;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:288]: This mean that we can read multiple times the same file which is a performance blackhole. Please use getParametersSet method

File: dynawo/sources/Modeler/Common/DYNDynamicData.h [hunk 7]
```diff
@@ -22,11 +22,13 @@
 #include <set>
 #include <string>
 #include <vector>
-
 #include <boost/shared_ptr.hpp>
 #include <boost/unordered_map.hpp>
 #include <boost/noncopyable.hpp>
 
+#include "DYDBlackBoxModel.h"
+#include "PARParameter.h"
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.h:30]: SHould be before the system include

File: dynawo/sources/Modeler/Common/DYNModeler.cpp [hunk 8]
```diff
@@ -163,6 +163,8 @@ Modeler::initParamDescription(const shared_ptr<ModelDescription>& modelDescripti
         } else {
           throw DYNError(Error::MODELER, ParameterWrongTypeReference, *itRef);
         }
+      } else if (refOrigData == Reference::PAR) {
+        continue;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNModeler.cpp:167]: Please add a comment to explain that those were resolvd in DynamicData

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_ReferenceToParameterFromAnotherModel/IEEE14.par [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    <par type="DOUBLE" name="governor_PNom" value="228"/>
+    <reference name="generator_P0Pu" origData="IIDM" origName="p_pu" type="DOUBLE"/>
+    <reference name="generator_Q0Pu" origData="IIDM" origName="q_pu" type="DOUBLE"/>
+    <reference name="generator_U0Pu" origData="IIDM" origName="v_pu" type="DOUBLE"/>
+    <reference name="generator_UPhase0" origData="IIDM" origName="angle_pu" type="DOUBLE"/>
+  </set>
+  <set id="7">
+    <par type="INT" name="nbGen" value="5"/>
+    <reference name="weight_gen_0" type="DOUBLE" parId="2" origName="generator_SNom" origData="PAR"/>
+    <reference name="weight_gen_1" type="DOUBLE" parId="7" origName="weight_gen_1" origData="PAR" parFile="../IEEE14_BlackBoxModels/IEEE14.par" />
```
[the remark this answers is not in the document]
Comment (barondim) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_ReferenceToParameterFromAnotherModel/IEEE14.par:267]: In this case, what is the point to reference another PAR file ?

[the remark this answers is not in the document]
Comment (rosiereflo) [reply nrt/data/IEEE14/IEEE14_SyntaxExamples/IEEE14_ReferenceToParameterFromAnotherModel/IEEE14.par:267]: Please create a local file inside the nrt folder, we would like them to be standalone.
This feature will be used in DFL.

File: dynawo/sources/API/PAR/PARReference.cpp [hunk 10]
```diff
@@ -27,7 +27,14 @@ using std::string;
 
 namespace parameters {
 
-Reference::Reference(const string& name) : type_(""), name_(name), origName_(""), componentId_("") {}
+Reference::Reference(const string& name, OriginData origData) :
```
[the remark this answers is not in the document]
Comment (barondim) [reply dynawo/sources/API/PAR/PARReference.cpp:30]: We must choose if the reference has origData IIDM or PAR. Because we have to choose at the creation of the object, we have to initialize origData_ in the constructor. I think we may need to delete the setOrigData method as well.

File: dynawo/sources/API/PAR/PARReference.cpp (hunk 2 above)
Comment (barondim) [reply dynawo/sources/API/PAR/PARReference.cpp:31]: All these attributes can be accessed in the code. It may be dangerous to not initialize them all.

Comment (rosiereflo) [reply dynawo/sources/API/PAR/PARReference.cpp:31]: The default constructor of strings create an empty string so it is useless to use the copy constructor for an empty string

File: dynawo/sources/API/PAR/PARReference.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/PAR/PARReference.cpp:30]: I mean: why do we need to specify it in the constructor when before it was not required?

[the remark this answers is not in the document]
Comment (barondim) [reply dynawo/sources/API/PAR/PARReference.cpp:30]: Before, IIDM was the only choice.

[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/PAR/PARReference.cpp:30]: Ok, so in this case let's not put a default value to the origData parameter and make sure that the setter is not used

File: dynawo/sources/API/PAR/test/TestReference.cpp [hunk 11]
```diff
@@ -32,28 +31,41 @@ namespace parameters {
 //-----------------------------------------------------
 
 TEST(APIPARTest, Reference) {
-  shared_ptr<Reference> refencePar = ReferenceFactory::newReference("standardRef");
+  shared_ptr<Reference> refencePar = boost::shared_ptr<Reference>(new Reference("standardRef"));
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/test/TestReference.cpp:34]: Please use the factory

**Comment by barondim** [comment]
To be merged with https://github.com/dynawo/dynaflow-launcher/pull/619

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp [hunk 12]
```diff
@@ -274,6 +275,36 @@ DynamicData::getParametersSet(const string& modelId, const string& parFile, cons
...
+      std::string parFile;
+      if (parRef->getParFile().empty()) {
+        parFile = modelSet->getFilePath();
+      } else {
+        parFile = parRef->getParFile();
+      }
+      std::string parFilePath = canonical(parFile, rootDirectory_);
+      shared_ptr<ParametersSet> referencedParametersSet = getParametersSet(model->getId(), parFilePath, parRef->getParId());
+      shared_ptr<Parameter> referencedParameter = referencedParametersSet->getParameter(parRef->getOrigName());
+      if (parRef->getType() == "DOUBLE") {
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: This might segfault if the parameter does not exist

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/Common/DYNDynamicData.cpp (hunk 12 above)
Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: The type_ attribute is  default initialized by the Reference constructor so there is no undefined behaviour possible.

Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: I'm not sure it's possible to test DynamicData::resolveParReferences method since it's a private method.

Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: parRef is a pointer to Reference class. It comes from the references_ map which is filled in by Xerces.

> > The type_ attribute is default initialized by the Reference constructor so there is no undefined behaviour possible.
> 
> but parRef can be nullptr no?

Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: sorry I meant referencedParameter

Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNDynamicData.cpp:295]: If the parId is wrong, the ParametersSetNotFound exception is thrown.
If the file containing the referenced value doesn't exist, the parsing will fail.
If the referenced value doesn't exist, the ParameterNotFoundInSet exception is thrown.
If the reference and the referenced value have different types, the ParameterInvalidTypeRequested exception is thrown.

File: documentation/configuringDynawo/DynawoInputFiles.tex [hunk 13]
```diff
@@ -679,19 +679,32 @@ \subsubsection{Define sets of static references}
 
 \subsection{par file}
 
-\subsubsection{Set a reference to a value stored in the iidm file}
+\subsubsection{Set a reference to a value stored in the iidm file or a given par file}
 \label{sec:Set a reference to a value stored in the iidm file}
 
 If the parameter's value is stored in the iidm file, a reference to this value can be added in the par file as in the example bellow:
-\begin{lstlisting}[language=XML, morekeywords={reference}]
+\begin{lstlisting}[language=XML, morekeywords={iidmreference}]
```
Comment (rosiereflo) [inline documentation/configuringDynawo/DynawoInputFiles.tex:686]: Why this change?

File: documentation/configuringDynawo/DynawoInputFiles.tex [hunk 14]
```diff
@@ -679,19 +679,32 @@ \subsubsection{Define sets of static references}
...
   <par type="ParType" name="ParName" value="ParValue"/>
   ...
   <reference type="RefType" name="RefName" origData="IIDM" origName="NameIIDM" componentId="Model_id_in_iidm_File"/>
 </set>
 \end{lstlisting}
 
-When no componentId is given, it is assumed that the data is linked with the static component which id is given by the staticId defined in the dyd file.
+When no componentId is given, it is assumed that the data is linked with the static component which id is given by the staticId defined in the dyd file.  \\
+
+If the parameter's value is stored in the par file, a reference to this value can be added in the par file as in the example below:
```
Comment (rosiereflo) [inline documentation/configuringDynawo/DynawoInputFiles.tex:696]: I would propose "To use the value from another parameter, a parameter reference can be used as shown in the example below:"

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/Common/DYNModelDescription.h [hunk 15]
```diff
@@ -61,87 +61,87 @@ class ModelDescription {
    * @brief add static reference interface
    * @param staticRefInterface static reference interface
    */
-  inline void addStaticRefInterface(const boost::shared_ptr<StaticRefInterface>& staticRefInterface) {
```
[the remark this answers is not in the document]
Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNModelDescription.h:64]: Methods entirely defined inside a class are implicitly inline

File: documentation/configuringDynawo/DynawoInputFiles.tex (hunk 13 above)
Comment (barondim) [reply documentation/configuringDynawo/DynawoInputFiles.tex:686]: Because I added another lstlisting below (about PAR reference)

File: dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic [hunk 16]
```diff
@@ -278,4 +278,5 @@ ReferenceAlreadySet         =             reference %1% is already set
 ReferenceAlreadySetInMacroParameterSet =   reference %1% is already set in macroParameterSet %2%
 ReferenceNotFoundInSet      =             reference %1% is not found in set %2%
 ReferenceUnknownOriginData  =             origin data %1% is not allowed for references
+ReferenceToAnotherReference =             reference %1% points to another reference %2% in par set %3% in file %4%
```
[the remark this answers is not in the document]
Comment (barondim) [reply dynawo/sources/Common/Dictionaries/DYNError_en_GB.dic:281]: If the set or the file are empty, there will be an error during the XML import according to my tests

File: documentation/configuringDynawo/DynawoInputFiles.tex (hunk 13 above)
Comment (rosiereflo) [reply documentation/configuringDynawo/DynawoInputFiles.tex:686]: I think this is just to tell LaTeX that reference should be seen as a xml key word

Comment (barondim) [reply documentation/configuringDynawo/DynawoInputFiles.tex:686]: Ok so I put "reference" back

**Review by rosiereflo — DISMISSED** [review]

**Review by rosiereflo — APPROVED** [review]
